consensus/polybft/slashing/double_signing_tracker.go

```diff
@@ -0,0 +1,221 @@
+}
+
+// createSenderMsgsMap initializes senders message map for the given round
+func createSenderMsgsMap(round uint64, roundMsgs map[uint64]SenderMessagesMap) SenderMessagesMap {
+	sendersMsgs := make(SenderMessagesMap)
+	roundMsgs[round] = sendersMsgs
+
+	return sendersMsgs
+}
+
+type DoubleSigningTracker interface {
+	Handle(msg *ibftProto.Message)
+	GetEvidences(height uint64) []*DoubleSignEvidence
+	PruneMsgsUntil(height uint64)
+}
+
+type DoubleSigningTrackerImpl struct {
+	preprepare  *Messages
+	prepare     *Messages
+	commit      *Messages
+	roundChange *Messages
+
+	logger hclog.Logger
+}
+
+func NewDoubleSigningTracker(logger hclog.Logger) *DoubleSigningTrackerImpl {
+	return &DoubleSigningTrackerImpl{
+		logger:      logger,
+		preprepare:  &Messages{content: make(MessagesMap)},
+		prepare:     &Messages{content: make(MessagesMap)},
+		commit:      &Messages{content: make(MessagesMap)},
+		roundChange: &Messages{content: make(MessagesMap)},
+	}
+}
+
+// Handle is implementation of IBFTMessageHandler interface, which handles IBFT consensus messages
+func (t *DoubleSigningTrackerImpl) Handle(msg *ibftProto.Message) {
+	if err := t.validateMsg(msg); err != nil {
+		t.logger.Debug("[ERROR] invalid IBFT message retrieved", "error", err, "message", msg)
+
+		return
+	}
+
+	sender := types.BytesToAddress(msg.From)
+
+	msgMap := t.resolveMessagesStorage(msg.GetType())
+	msgMap.addMessage(msg.View, sender, msg)
+}
+
+// PruneMsgsUntil deletes all messages maps until the specified height
+func (t *DoubleSigningTrackerImpl) PruneMsgsUntil(height uint64) {
+	for _, msgType := range ibftProto.MessageType_value {
+		msgs := t.resolveMessagesStorage(ibftProto.MessageType(msgType))
+		if msgs == nil {
+			continue
+		}
+
+		msgs.mux.Lock()
+		for msgHeight := range msgs.content {
+			if msgHeight < height {
+				delete(msgs.content, msgHeight)
+			}
+		}
+		msgs.mux.Unlock()
+	}
+}
+
+// GetEvidences returns double signing evidences for the given height
+func (t *DoubleSigningTrackerImpl) GetEvidences(height uint64) []*DoubleSignEvidence {
+	evidences := []*DoubleSignEvidence{}
+
+	for _, msgType := range ibftProto.MessageType_value {
+		msgs := t.resolveMessagesStorage(ibftProto.MessageType(msgType))
+		if msgs == nil {
+			continue
+		}
+
+		roundMsgs, ok := msgs.content[height]
+		if !ok {
+			continue
+		}
+
+		for round, senderMsgs := range roundMsgs {
+			for address, msgs := range senderMsgs {
+				if len(msgs) <= 1 {
+					continue
+				}
+
+				var evidence *DoubleSignEvidence
+
+				firstMsg := msgs[0]
+				for _, msg := range msgs[1:] {
+					if !bytes.Equal(firstMsg.Signature, msg.Signature) {
+						if evidence == nil {
+							evidence = newDoubleSignEvidence(address, round, []*ibftProto.Message{})
+							evidences = append(evidences, evidence)
+						}
+
+						evidence.messages = append(evidence.messages, msg)
+					}
+				}
+
+				evidence.messages = append([]*ibftProto.Message{firstMsg}, evidence.messages...)
```

Comment:
Evidence could be nil here if some byzantine validator submitted the same message and signature two or more times.
I think we should also filter such messages in Handle() to protect from possible spam attacks that could (if someone is persistent) crash the node when it runs out of memory. If we filter them in Handle() then we can move evidence instantiation before the inner most loop. Does it make sense?

Reply:
Fixed https://github.com/0xPolygon/polygon-edge/pull/1756/commits/67879095f8b35a6828e5210a417f012af4af6949